src/common/pkcs11/pkcs11.cpp

```diff
@@ -0,0 +1,1059 @@
+        return ErrorEnum::eInvalidArgument;
+    }
+
+    if (dst.MaxSize() < types.Size()) {
+        return ErrorEnum::eNoMemory;
+    }
+
+    dst.Clear();
+
+    for (size_t i = 0; i < types.Size(); ++i) {
+        CK_ATTRIBUTE tmp;
+
+        tmp.type = types[i];
+        tmp.pValue = static_cast<void*>(values[i].Get());
+        tmp.ulValueLen = values[i].MaxSize();
+
+        dst.PushBack(tmp);
+    }
+
+    return ErrorEnum::eNone;
+}
+
+Error GetAttributesValues(Array<CK_ATTRIBUTE>& src, Array<Array<uint8_t>>& values)
+{
+    if (values.MaxSize() < src.Size()) {
+        return ErrorEnum::eNoMemory;
+    }
+
+    values.Clear();
+
+    for (const auto& attr : src) {
+        if (attr.ulValueLen == CK_UNAVAILABLE_INFORMATION) {
+            return ErrorEnum::eInvalidArgument;
+        }
+
+        auto tmp = Array<uint8_t>(static_cast<uint8_t*>(attr.pValue), attr.ulValueLen);
+        values.PushBack(tmp);
+    }
+
+    return ErrorEnum::eNone;
+}
+
+Array<uint8_t> ConvertToAttributeValue(const String& val)
+{
+    return {const_cast<uint8_t*>(reinterpret_cast<const uint8_t*>(val.Get())), val.Size()};
+}
+
+template <typename T>
+Array<uint8_t> ConvertToAttributeValue(T& val)
+{
+    return {reinterpret_cast<uint8_t*>(&val), sizeof(val)};
+}
+
+/***********************************************************************************************************************
+ * LibraryContext
+ **********************************************************************************************************************/
+
+LibraryContext::LibraryContext(void* handle)
+    : mHandle(handle)
+{
+}
+
+Error LibraryContext::Init()
+{
+    CK_C_GetFunctionList getFuncList = reinterpret_cast<CK_C_GetFunctionList>(dlsym(mHandle, "C_GetFunctionList"));
+
+    if (getFuncList == nullptr) {
+        LOG_ERR() << "Can't find C_GetFunctionList: dlsym err = " << dlerror();
+
+        return ErrorEnum::eFailed;
+    }
+
+    CK_RV rv = getFuncList(&mFunctionList);
+    if (rv != CKR_OK) {
+        LOG_ERR() << "C_GetFunctionList failed: err = " << rv;
+
+        return ErrorEnum::eFailed;
+    }
+
+    if (mFunctionList == nullptr) {
+        LOG_ERR() << "C_GetFunctionList returned NULL";
+
+        return ErrorEnum::eFailed;
+    }
+
+    rv = mFunctionList->C_Initialize(nullptr);
+    if (rv != CKR_OK) {
+        LOG_ERR() << "C_Initialize failed: err = " << rv;
+
+        return ErrorEnum::eFailed;
+    }
+
+    return ErrorEnum::eNone;
+}
+
+Error LibraryContext::InitToken(SlotID slotID, const String& pin, const String& label)
+{
+    if (!mFunctionList || !mFunctionList->C_InitToken) {
+        LOG_ERR() << "C_InitToken failed. Library is not initialized.";
+
+        return ErrorEnum::eFailed;
+    }
+
+    CK_UTF8CHAR pkcsLabel[cLabelLen];
+
+    auto err = ConvertToPKCS11String(label, pkcsLabel);
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    CK_RV rv = mFunctionList->C_InitToken(slotID, (CK_UTF8CHAR_PTR)pin.CStr(), pin.Size(), pkcsLabel);
```

Comment:
can we use here reinterprate_cast here ?

Reply:
basically that's a combination of reinterpret & const cast, as in some cases even readonly value require non-const buffers. So I used c-style cast instead. 
Replaced with a function.